File: nast.c
/**
Fairing Sizer - Show PLF Dimensions on Notebook Paper Sized Drawing
Copyright (C) 2019  Andrew C. Suttles

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

Andrew Suttles - andrew.suttles@gmail.com
**/
#ifdef ___LINKER_INFO
; File: "nast.c", produced by Gambit-C v4.6.0
(
406000
" nast"
(" nast")
(
)
(
)
(
" nast"
)
(
"test-scheme"
)
(
"display"
"newline"
)
 #f
)
#else
#define ___VERSION 406000
#define ___MODULE_NAME " nast"
#define ___LINKER_ID ____20_nast
#define ___MH_PROC ___H__20_nast
#define ___SCRIPT_LINE 0
#define ___GLO_COUNT 4
#define ___SUP_COUNT 2
#define ___SUB_COUNT 1
#define ___LBL_COUNT 7
#include "gambit.h"

___NEED_GLO(___G__20_nast)
___NEED_GLO(___G_display)
___NEED_GLO(___G_newline)
___NEED_GLO(___G_test_2d_scheme)

___BEGIN_GLO
___DEF_GLO(0," nast")
___DEF_GLO(1,"test-scheme")
___DEF_GLO(2,"display")
___DEF_GLO(3,"newline")
___END_GLO

___DEF_SUB_STR(___X0,11)
               ___STR8(104,101,108,108,111,32,119,111)
               ___STR3(114,108,100)

___BEGIN_SUB
 ___DEF_SUB(___X0)
___END_SUB



#undef ___MD_ALL
#define ___MD_ALL ___D_FP ___D_R0 ___D_R1 ___D_R4
#undef ___MR_ALL
#define ___MR_ALL ___R_FP ___R_R0 ___R_R1 ___R_R4
#undef ___MW_ALL
#define ___MW_ALL ___W_FP ___W_R0 ___W_R1 ___W_R4
___BEGIN_M_COD
___BEGIN_M_HLBL
___DEF_M_HLBL_INTRO
___DEF_M_HLBL(___L0__20_nast)
___DEF_M_HLBL_INTRO
___DEF_M_HLBL(___L0_test_2d_scheme)
___DEF_M_HLBL(___L1_test_2d_scheme)
___DEF_M_HLBL(___L2_test_2d_scheme)
___DEF_M_HLBL(___L3_test_2d_scheme)
___END_M_HLBL

___BEGIN_M_SW

#undef ___PH_PROC
#define ___PH_PROC ___H__20_nast
#undef ___PH_LBL0
#define ___PH_LBL0 1
#undef ___PD_ALL
#define ___PD_ALL ___D_R0 ___D_R1
#undef ___PR_ALL
#define ___PR_ALL ___R_R0 ___R_R1
#undef ___PW_ALL
#define ___PW_ALL ___W_R1
___BEGIN_P_COD
___BEGIN_P_HLBL
___DEF_P_HLBL_INTRO
___DEF_P_HLBL(___L0__20_nast)
___END_P_HLBL
___BEGIN_P_SW
___DEF_SLBL(0,___L0__20_nast)
   ___IF_NARGS_EQ(0,___NOTHING)
   ___WRONG_NARGS(0,0,0,0)
___DEF_GLBL(___L__20_nast)
   ___SET_GLO(1,___G_test_2d_scheme,___PRC(3))
   ___SET_R1(___VOID)
   ___JUMPPRM(___NOTHING,___R0)
___END_P_SW
___END_P_COD

#undef ___PH_PROC
#define ___PH_PROC ___H_test_2d_scheme
#undef ___PH_LBL0
#define ___PH_LBL0 3
#undef ___PD_ALL
#define ___PD_ALL ___D_FP ___D_R0 ___D_R1 ___D_R4
#undef ___PR_ALL
#define ___PR_ALL ___R_FP ___R_R0 ___R_R1 ___R_R4
#undef ___PW_ALL
#define ___PW_ALL ___W_FP ___W_R0 ___W_R1 ___W_R4
___BEGIN_P_COD
___BEGIN_P_HLBL
___DEF_P_HLBL_INTRO
___DEF_P_HLBL(___L0_test_2d_scheme)
___DEF_P_HLBL(___L1_test_2d_scheme)
___DEF_P_HLBL(___L2_test_2d_scheme)
___DEF_P_HLBL(___L3_test_2d_scheme)
___END_P_HLBL
___BEGIN_P_SW
___DEF_SLBL(0,___L0_test_2d_scheme)
   ___IF_NARGS_EQ(0,___NOTHING)
   ___WRONG_NARGS(0,0,0,0)
___DEF_GLBL(___L_test_2d_scheme)
   ___SET_STK(1,___R0)
   ___SET_R1(___SUB(0))
   ___SET_R0(___LBL(2))
   ___ADJFP(4)
   ___POLL(1)
___DEF_SLBL(1,___L1_test_2d_scheme)
   ___JUMPGLOSAFE(___SET_NARGS(1),2,___G_display)
___DEF_SLBL(2,___L2_test_2d_scheme)
   ___SET_R0(___STK(-3))
   ___POLL(3)
___DEF_SLBL(3,___L3_test_2d_scheme)
   ___ADJFP(-4)
   ___JUMPGLOSAFE(___SET_NARGS(0),3,___G_newline)
___END_P_SW
___END_P_COD

___END_M_SW
___END_M_COD

___BEGIN_LBL
 ___DEF_LBL_INTRO(___H__20_nast," nast",___REF_FAL,1,0)
,___DEF_LBL_PROC(___H__20_nast,0,0)
,___DEF_LBL_INTRO(___H_test_2d_scheme,0,___REF_FAL,4,0)
,___DEF_LBL_PROC(___H_test_2d_scheme,0,0)
,___DEF_LBL_RET(___H_test_2d_scheme,___IFD(___RETI,4,0,0x3f1L))
,___DEF_LBL_RET(___H_test_2d_scheme,___IFD(___RETN,1,0,0x1L))
,___DEF_LBL_RET(___H_test_2d_scheme,___IFD(___RETI,4,4,0x3f0L))
___END_LBL

___BEGIN_MOD1
___DEF_PRM(0,___G__20_nast,1)
___END_MOD1

___BEGIN_MOD2
___END_MOD2

#endif

File: nast.lua
-- Write a NASTRAN BDF for Payload Fairing

-- Fairing Sizer - Show PLF Dimensions on Notebook Paper Sized Drawing
-- Copyright (C) 2019  Andrew C. Suttles
-- 
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
-- 
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
-- 
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
-- 
-- Andrew Suttles - andrew.suttles@gmail.com


local _deg2rad = math.pi / 180.0	  

-- GLOBALS for GEOMETRY

local _L			-- Overall Length
local _Lo			-- Length Ogive Section
local _Lb			-- Length Barrel Section

local _Db			-- Diameter Barrel

local _Rn			-- Nose Radius
local _Xt			-- Nose Xt (Tangency)

local _S 			-- Grid Spacing


-- ------------------------------------------------------------------------------
--			   Helper Functions
-- ------------------------------------------------------------------------------

local function write_blank_line(f,n)
   for i = 1,n do
      f:write( "$\n" )
   end
end

local function write_ruler(f)
   f:write( "$--01--][--02--][--03--][--04--][--05--][--06--][--07--][--08--][--09--][--10--]\n" )
end

local function write_cord2c(f)
   f:write( "CORD2C         1               0       0       0       0       0       1\n" )
   f:write( "               1       0       0\n" )
end

local function write_grid_description(f)

   write_blank_line(f,1)
   f:write( "$  GRID IDs:  XXYYYZZZ\n" )
   f:write( "$             XX       = Section ID( Barrel=10, Ogive=20, Nose=30\n" )
   f:write( "$                YYY    = Ring Number for this section ( < 500  )\n" )
   f:write( "$                   ZZZ = Grid Number for this ring (0 - 359 )\n" ) 

end

-- Write a circumferential ring of grids
-- @Out: None
-- @Params:
--     f    = file handle
--     elID = fairing segment ID
--     i    = ring number
--     z    = z Coordinate
--     R    = R Coordinate
local function write_grid_ring( f, elID, i, z, R )

   --Grid ID    Section ID           Ring ID
   local id = ( 1000000 * elID ) + ( 1000 * i )

   for j = 0,359 do		       -- 360 degrees

      f:write( string.format( "GRID    %-8d%8d%8.2f%8.5f%8.2f\n", 
			      id + j, 1, R, ( _deg2rad * j), z ))
   end
end

-- ------------------------------------------------------------------------------
--			   Write Barrel BDF
-- ------------------------------------------------------------------------------

local function write_barrel_grids( f ) 
   

   local Rb = ( _Db / 2.0 )		  -- radius of barrel diameter
   _S        = Rb * _deg2rad		  -- dist between grids
                                          -- calc num vertical rings 
   local n  = math.min( math.floor( _Lb / _S ), 500 )

   -- Lower and Middle Rings
   for i = 0,n do
      local z = ( _Lb / n ) * i
      write_grid_ring( f, 10, i, z, Rb )
   end
   
   -- Top Ring
   write_grid_ring( f, 10, n+1, _Lb, Rb )
end


local function write_barrel_bdf()

   local f = io.open( "barrel.bdf", "w" ) -- NASTRAN bdf for barrel section

   write_grid_description(f)
   write_blank_line(f,1)
   write_ruler(f)
   write_barrel_grids( f )

   f:close()
end

-- ------------------------------------------------------------------------------
--			   Write Ogive BDF
-- ------------------------------------------------------------------------------

local function write_ogive_grids( f )

   local n  = math.min( math.floor( _Lo / _S ), 500 )

   -- Lower and Middle Rings
   for i = 0,n do
      local z = ( _Lo / n ) * i
      write_grid_ring( f, 20, i, _Lb + z, ogiveY( z / 12.0 ))
   end

   -- Top Ring
   write_grid_ring( f, 20, n+1, _Lb + _Lo, ogiveY( _Lo / 12.0 ))
end


local function write_ogive_bdf()

   local f = io.open( "ogive.bdf", "w" ) -- NASTRAN bdf for ogive section

   write_grid_description(f)
   write_blank_line(f,1)
   write_ruler(f)
   write_ogive_grids( f )

   f:close()
end


-- ------------------------------------------------------------------------------
--			   Write Ogive BDF
-- ------------------------------------------------------------------------------

local function write_nose_grids( f )

   local i = 0
   local z0 = _Lb + _Lo

   -- Lower and Middle Rings
   for x = _Xt,_Rn,_S do
      local z = z0 + (x - _Xt)
      local h = _Rn - x		-- dist from x to Rn
      write_grid_ring( f, 30, i, z, math.sqrt(( 2.0 * h * _Rn ) - ( h * h )))
      i = i + 1
   end

   -- Top Grid
   f:write( string.format( "GRID    %-8d%8d%8.2f%8.5f%8.2f\n", 
			   30000000 + (1000 * i),
			   1, 0.0, 0.0, _L ))
end

local function write_nose_bdf()

   local f = io.open( "nose.bdf", "w" ) -- NASTRAN bdf for ogive section

   write_grid_description(f)
   write_blank_line(f,1)
   write_ruler(f)
   write_nose_grids( f )

   f:close()

end

-- ------------------------------------------------------------------------------
--			      Write BDF
-- ------------------------------------------------------------------------------

local function set_global_state()

   _L  = getLength()
   _Lo = getOgiveLength()
   _Lb = getBarrelLength()
   _Db = getBaseDiameter()
   _Rn = getNoseRadius()
   _Xt = getNoseCapXt()

end

function write_deck()

   set_global_state()		-- Load Fairing Dimensions
   
   local f = io.open( "fairing.bdf", "w" ) -- Create NASTRAN bdf 

   f:write( "BEGIN BULK\n" )
   write_blank_line(f,2)

   -- Coord System
   write_cord2c(f)
   write_blank_line(f, 2)

   -- Barrel Section
   f:write( "INCLUDE 'barrel.bdf'\n" )
   write_barrel_bdf( Db, Lb )
   
   -- Ogive Section
   f:write( "INCLUDE 'ogive.bdf'\n" )
   write_ogive_bdf()

   -- Nose Cap Section
   f:write( "INCLUDE 'nose.bdf'\n" )
   write_nose_bdf()

   f:close()
end


